### PR TITLE
Fixup bugs and refactor ModelRegions

### PR DESCRIPTION
**Pull request by bgeltz**
Address issues that have become apparent based on the changes that went in with PR #1757.

### CODE REVIEW

**Review by bgeltz — COMMENTED** [review]
Move warmup loop into it's own method and get it out of big_o().

File: src/Profile.cpp [hunk 1]
```diff
@@ -412,11 +412,19 @@ namespace geopm
...
-            throw Exception("ProfileImp:region() multiple region hints set and only 1 at a time is supported.",
+        geopm::check_hint(hint);
+        uint64_t result = m_table->key(region_name);
+
+#ifdef GEOPM_DEBUG
+        if (m_region_ids.find(result) == m_region_ids.end()) {
+            m_region_ids.insert(result);
+        }
+        else {
+            throw Exception("Profile::region(): Region '" + region_name + "' already created.",
```
Comment (bgeltz) [inline src/Profile.cpp]: It's legal to call prog_region twice with the same region_name + hint combo.

File: src/Profile.hpp [hunk 2]
```diff
@@ -314,6 +315,9 @@ namespace geopm
             double m_overhead_time_startup;
             double m_overhead_time_shutdown;
 
+            /// @brief The list of known region identifiers.
+            std::unordered_set<uint64_t> m_region_ids;
```
Comment (bgeltz) [inline src/Profile.hpp]: Wrap this in ifdef debug too.

File: test/ProfileTest.cpp [hunk 3]
```diff
@@ -152,12 +156,18 @@ TEST_F(ProfileTest, enter_exit)
...
+    uint64_t usr_region_id = m_profile->region(usr_name, usr_hint);
+    uint64_t usr_hash = geopm_region_id_hash(usr_region_id);
+
+    EXPECT_CALL(*m_table, key(_)).WillOnce(Return(0x5678));
+    std::string mpi_name = "mpi_test_region";
     uint64_t mpi_hint = GEOPM_REGION_HINT_NETWORK;
-    uint64_t mpi_region_id = mpi_hint | mpi_hash;
+    uint64_t mpi_region_id = m_profile->region(mpi_name, mpi_hint);
+    uint64_t mpi_hash = geopm_region_id_hash(mpi_region_id);
+
```
Comment (bgeltz) [inline test/ProfileTest.cpp:169]: enter(region_id) instead of or'ing things together.

File: test/ProfileTestIntegration.cpp [hunk 4]
```diff
@@ -246,12 +251,18 @@ TEST_F(ProfileTestIntegration, enter_exit_short)
...
+    uint64_t usr_region_id = m_profile->region(usr_name, usr_hint);
+    uint64_t usr_hash = geopm_region_id_hash(usr_region_id);
+
+    EXPECT_CALL(*m_table, key(_)).WillOnce(Return(0x5678));
+    std::string mpi_name = "mpi_test_region";
     uint64_t mpi_hint = GEOPM_REGION_HINT_NETWORK;
-    uint64_t mpi_region_id = mpi_hint | mpi_hash;
+    uint64_t mpi_region_id = m_profile->region(mpi_name, mpi_hint);
+    uint64_t mpi_hash = geopm_region_id_hash(mpi_region_id);
+
```
Comment (bgeltz) [inline test/ProfileTestIntegration.cpp:264]: enter(region_id)

File: src/ModelRegion.cpp [hunk 5]
```diff
@@ -172,15 +172,27 @@ namespace geopm
...
     void ModelRegion::region_enter(void)
     {
+        int err = 0;
         if (!m_do_unmarked) {
-            (void)geopm_prof_enter(m_region_id);
+            err = geopm_prof_enter(m_region_id);
+            if (err != 0) {
+                throw Exception("ModelRegion::region_enter(): geopm_prof_enter() error on region_id: '" +
+                                geopm::string_format_hex(m_region_id)  + "'",
+                                GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
```
Comment (cmcantalupo) [inline src/ModelRegion.cpp]: GEOPM_ERROR_RUNTIME -> err?

File: src/ModelRegion.cpp [hunk 6]
```diff
@@ -172,15 +172,27 @@ namespace geopm
...
     void ModelRegion::region_exit(void)
     {
+        int err = 0;
         if (!m_do_unmarked) {
-            (void)geopm_prof_exit(m_region_id);
+            err = geopm_prof_exit(m_region_id);
+            if (err != 0) {
+                throw Exception("ModelRegion::region_exit(): geopm_prof_exit() error on region_id: '" +
+                                geopm::string_format_hex(m_region_id)  + "'",
+                                GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
```
Comment (cmcantalupo) [inline src/ModelRegion.cpp]: GEOPM_ERROR_RUNTIME -> err?

File: src/Profile.cpp [hunk 7]
```diff
@@ -412,11 +412,15 @@ namespace geopm
...
 
-        if (hint && ((hint & (hint - 1)) != 0)) {   /// power of 2 check
-            throw Exception("ProfileImp:region() multiple region hints set and only 1 at a time is supported.",
-                            GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
-        }
+        geopm::check_hint(hint);
         uint64_t result = m_table->key(region_name);
+
+#ifdef GEOPM_DEBUG
+        if (m_region_ids.find(result) == m_region_ids.end()) {
```
Comment (cmcantalupo) [inline src/Profile.cpp]: Use emplace(), rather than the find/insert combo.

https://en.cppreference.com/w/cpp/container/set/emplace

**Review by cmcantalupo — COMMENTED** [review]
Just a couple of nits.  Looks better!

**Review by cmcantalupo — APPROVED** [review]

**Pull request merged**